hivemind/optim/averaged.py

```diff
@@ -7,61 +9,114 @@
 
 
+@dataclass(frozen=False)
+class DecentralizedState:
+    max_epoch: int = 0
+    total_steps: int = 0
+
 class DecentralizedOptimizer(DecentralizedOptimizerBase):
     """
     A simple optimizer that trains a shared model by averaging with peers in variety of ways. Supports
     parameter/gradient averaging and syncing adaptive learning rates or any other internal statistics of optimizer.
 
     :param opt: a pytorch optimizer configured to update model parameters.
     :param dht: a running hivemind DHT daemon connected to other peers
+    :param prefix: all DHT keys that point to optimization metadata will have this prefix
+    :param target_group_size: maximum group size for averaging (see DecentralizedAverager)
     :param average_parameters: whether to average model parameters
     :param average_gradients: whether to average gradients
+    :param max_allowed_epoch_difference: if max_epoch has difference with local_epoch mote than that, we download state
+      from other peer.
+    :param total_steps_in_epoch: how many total steps must be to increase local_epoch by one
     :param average_opt_statistics: if specified, average optimizer states with corresponding names in state_dict
+    :param scheduler_cls: lambda with opt in argument which returns learning rate scheduler
     :param averaging_steps_period: performs averaging after this many optimizer steps
     :param averaging_time_period: if specified, optimizer will attempt to average weights at regular intervals of this
       many seconds. (averaging step will only occur if the optimizer ran `averaging_steps_period` steps in that interval)
-    :param prefix: all DHT keys that point to optimization metadata will have this prefix
-    :param target_group_size: maximum group size for averaging (see DecentralizedAverager)
+    report_progress_expiration
     :param timeout: if DecentralizedAverager step is unable to form group in this many seconds, cancel step
+    :param verbose: verbose info
     :param kwargs: additional parameters passed to TrainingAverager
     :note: if you're using an optimizer with adaptive learning rates (such as Adam), make sure to specify
       necessary fields' names in `average_opt_statistics`. Otherwise you may encounter poor convergence.
     :note: the base optimizer cannot add param groups after the DecentralizedOptimizer is created
     """
 
     def __init__(self, opt: torch.optim.Optimizer, dht: DHT, *, prefix: str, target_group_size: int,
-                 average_parameters: bool, average_gradients: bool, average_opt_statistics: Sequence[str] = (),
-                 averaging_steps_period: int = 1, averaging_time_period: float = 0,
-                 timeout: Optional[float] = None, verbose: bool = False, **kwargs):
+                 average_parameters: bool, average_gradients: bool, max_allowed_epoch_difference: int = 1,
+                 total_steps_in_epoch: int = 1000, average_opt_statistics: Sequence[str] = (),
```

Comment:
I'd keep the existing order, so that average_opt_statistics is next to average_gradients